### PR TITLE
When high qps, async operation code stack may too deep

### PR DESCRIPTION
**Pull request by semistone**

### CODE REVIEW

**Comment by mrniko** [comment]
Thanks for suggested changes. Can you share the example of stack you are trying to avoid?

**Comment by semistone** [comment]
[stack.gz](https://github.com/user-attachments/files/15951229/stack.gz)


it's related to issue
https://github.com/redisson/redisson/issues/5971

**Comment by semistone** [comment]
it's simple POC about  reduce call stack in AsyncSemaphore
```java
    private void tryRun() {
        while (true) {
            if (counter.decrementAndGet() >= 0) {
                CompletableFuture<Lock> future = listeners.poll();
                if (future == null) {
                    counter.incrementAndGet();
                    return;
                }

                if (!future.isDone()) {
                    // release only once by lock object
                    AtomicBoolean once = new AtomicBoolean();
                    AtomicBoolean tryNext = new AtomicBoolean();
                    long currentThreadId = Thread.currentThread().getId();
                    if (future.complete(() -> { <==== return release runnable 
                        if (once.compareAndSet(false, true)) {
                            // if in the same thread and tryNext is false means already release in the same thread.
                            release(Thread.currentThread().getId() != currentThreadId || tryNext.compareAndSet(false, true));
                        }
                    })) {
                        if (tryNext.compareAndSet(false, true)) { 
                            continue; // continue to tryRun  if not release yet.
                        } else {
                            return;
                        }
                    }
                }
            }

            if (counter.incrementAndGet() <= 0) {
                return;
            }
        }
    
```

I didn't test and optimism it yet.

File: redisson/src/main/java/org/redisson/connection/ConnectionsHolder.java [hunk 1]
```diff
@@ -202,9 +204,12 @@ public CompletableFuture<T> acquireConnection(RedisCommand<?> command) {
         CompletableFuture<T> result = new CompletableFuture<>();
 
         CompletableFuture<Void> f = acquireConnection();
-        f.thenAccept(r -> {
-            connectTo(result, command);
-        });
+        // prevent deep stack, switch thread per 100 requests.
+        if (count.incrementAndGet() % 100 == 0) {
+            f.thenAcceptAsync(r -> connectTo(result, command));
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/connection/ConnectionsHolder.java:209]: you need to define `serviceManager.getGroup()` as an executor.

Comment (mrniko) [reply redisson/src/main/java/org/redisson/connection/ConnectionsHolder.java:209]: serviceManager.getGroup().submit(() -> {
                releaseConnection();
});

now can be replaced with only releaseConnection();

**Review by mrniko — CHANGES_REQUESTED** [review]

**Comment by semistone** [comment]
I change to always switch thread during release as your suggestion.
I checked when that error happen, it those StackOverflowError and it's not Exception but throwable.
and the stack size is also depend on how user's implementation, it could happen after very few recursive loop. 
so switch thread in every release is  is better.

**Comment by semistone** [comment]
or switch in 

```java
RedisExecutor.sendCommand
            if (connectionManager.getServiceManager().getConfig().getMasterConnectionPoolSize() < 10
                    && !command.isBlockingCommand()) {
                release(connection); <=== here
            }
```

**Comment by mrniko** [comment]
Thanks for your assistance. I have fixed it in https://github.com/redisson/redisson/commit/3c1c90f60349e5f13742641819f9619a59101b08

**Pull request closed without merging**